pkg/client/events_tracking_test.go

```diff
@@ -0,0 +1,361 @@
+package client
+
+import (
+	"encoding/hex"
+	"reflect"
+	"sync"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+func TestRequestedSignersTrackAdd(t *testing.T) {
+	keepAddress1 := common.BytesToAddress([]byte{1})
+	keepAddress2 := common.BytesToAddress([]byte{2})
+
+	rs := &requestedSignersTrack{
+		data:  make(map[string]bool),
+		mutex: &sync.Mutex{},
+	}
+
+	if !rs.add(keepAddress1) {
+		t.Error("unexpected failure")
+	}
+
+	if ok := rs.data[keepAddress1.String()]; !ok {
+		t.Errorf(
+			"unexpected value for keep [%s]\nexpected: [%v]\nactual:   [%v]",
+			keepAddress1.String(),
+			true,
+			rs.data[keepAddress1.String()],
+		)
+	}
```

Comment:
Do we want to check it? I think we are going to deep inside the implementation, we should limit ourselves to the result returned by `add`.

Reply:
I'm 50-50. How do we ensure that `add` actually introduced changes that we expected? Maybe we could check expected map similiary as we do fo signatures?
```go
expectedData := map[string]bool{
	keepAddress1.String(): true,
	keepAddress2.String(): true,
}

if ok := !rs.add(keepAddress1); !ok {
	t.Error("unexpected failure")
}

if ok := rs.add(keepAddress2); !ok {
	t.Error("unexpected failure")
}

if !reflect.DeepEqual(expectedData, rs.data) {
	t.Errorf(
		"unexpected map content\nexpected: [%v]\nactual:   [%v]",
		expectedData,
		rs.data,
	)
}
```

Reply:
I think we shouldn't care about what's happening inside in test but what are the outcomes and how the API behaves. All we care is what `add` returns. So for signing:
```
func TestRequestedSignersTrackAdd(t *testing.T) {
	keepAddress1 := common.BytesToAddress([]byte{1})
	keepAddress2 := common.BytesToAddress([]byte{2})

	rs := &requestedSignersTrack{
		data:  make(map[string]bool),
		mutex: &sync.Mutex{},
	}

	if !rs.add(keepAddress1) {
		t.Error("event wasn't emitted before; should be added successfully")
	}

	if !rs.add(keepAddress2) {
		t.Error("event wasn't emitted before; should be added successfully")
	}
}

func TestRequestedSignersTrackAdd_Duplicate(t *testing.T) {
	keepAddress := common.BytesToAddress([]byte{1})

	rs := &requestedSignersTrack{
		data:  make(map[string]bool),
		mutex: &sync.Mutex{},
	}

	if !rs.add(keepAddress) {
		t.Error("event wasn't emitted before; should be added successfully")
	}

	if rs.add(keepAddress) {
		t.Error("event was emitted before; it should not be added")
	}
}

func TestRequestedSignersTrackRemove(t *testing.T) {
	keepAddress := common.BytesToAddress([]byte{1})

	rs := &requestedSignersTrack{
		data:  make(map[string]bool),
		mutex: &sync.Mutex{},
	}

	if !rs.add(keepAddress) {
		t.Error("event wasn't emitted before; should be added successfully")
	}

	rs.remove(keepAddress)

	if rs.add(keepAddress) {
		t.Error("event was removed from tracking; should be added successfully")
	}
}

func TestRequestedSignersTrackRemove_WhenEmpty(t *testing.T) {
	keepAddress := common.BytesToAddress([]byte{1})

	rs := &requestedSignersTrack{
		data:  make(map[string]bool),
		mutex: &sync.Mutex{},
	}

	rs.remove(keepAddress)

	if !rs.add(keepAddress) {
		t.Error("event wasn't emitted before; should be added successfully")
	}
}
```

Reply:
Okey!

Reply:
OK, let me update the rest of the tests as I am already 50% there.